README.md
# NAME

Protocol::Star::Linemode - Generates a formatted byte string for Star POS printers

# SYNOPSIS

    use Protocol::Star::Linemode;

    my $p = Protocol::Star::Linemode->new;
    $p->set_emphasized_printing;
    $p->text("Hello world");
    $p->cancel_emphasized_printing;

    my $formatted_output = $p->result;
    # Send $formatted_output to printer

# Converting from 0.1.2 to 1.0.0

- Create a [Protocol::Star::Linemode](http://search.cpan.org/perldoc?Protocol::Star::Linemode) object instead of the [Protocol::Star::Linemode::Generated](http://search.cpan.org/perldoc?Protocol::Star::Linemode::Generated) object

# AUTHOR

Peter Stuifzand <peter@stuifzand.eu>

# COPYRIGHT

Copyright 2013 - Peter Stuifzand

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# SEE ALSO


Changes
Changes for Protocol-Star-Linemode

{{$NEXT}}

1.0.0  2014-01-27 11:05:05 CET
		- Use role instead of sub class for generated code.

0.1.2  2013-08-08 11:12:54 CEST
		- Fixes to docs

0.1.1  2013-08-08 10:25:01 CEST
		- Converted to Milla


lib/Protocol/Star/Linemode/Generated.pm
# Generated file - do not modify
# Copyright (c) 2013 Peter Stuifzand
# Copyright (c) 2013 Other contributors as noted in the AUTHORS file
#
# Protocol::Star::Linemode::Generated is part of Protocol::Star::Linemode
#
# Protocol::Star::Linemode is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation; either version 3 of the License,
# or (at your option) any later version.
#
# Protocol::Star::Linemode is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser
# General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
package Protocol::Star::Linemode::Generated;
use Moo::Role;

sub command_initialization {
    my ($self, ) = @_;
    $self->append_pack("CC", 0x1B, 0x40, );
    return;
}

sub select_font {
    my ($self, $arg0) = @_;
    $self->append_pack("CCCC", 0x1B, 0x1E, 0x46, $arg0);
    return;
}

sub select_code_page {
    my ($self, $arg0) = @_;
    $self->append_pack("CCCC", 0x1B, 0x1D, 0x74, $arg0);
    return;
}

sub set_slash_zero {
    my ($self, $arg0) = @_;
    $self->append_pack("CCC", 0x1B, 0x27, $arg0);
    return;
}

sub specify_international_charset {
    my ($self, $arg0) = @_;
    $self->append_pack("CCC", 0x1B, 0x52, $arg0);
    return;
}

sub specify_12_dot_pitch {
    my ($self, ) = @_;
    $self->append_pack("CC", 0x1B, 0x4D, );
    return;
}

sub specify_15_dot_pitch {
    my ($self, ) = @_;
    $self->append_pack("CC", 0x1B, 0x50, );
    return;
}

sub specify_16_dot_pitch {
    my ($self, ) = @_;
    $self->append_pack("CC", 0x1B, 0x3A, );
    return;
}

sub specify_14_dot_pitch {
    my ($self, ) = @_;
    $self->append_pack("CC", 0x1B, 0x67, );
    return;
}

sub set_expanded_width_height {
    my ($self, $arg0, $arg1) = @_;
    $self->append_pack("CCCC", 0x1B, 0x69, $arg0, $arg1);
    return;
}

sub set_expanded_width {
    my ($self, $arg0) = @_;
    $self->append_pack("CCC", 0x1B, 0x57, $arg0);
    return;
}

sub set_expanded_height {
    my ($self, $arg0) = @_;
    $self->append_pack("CCC", 0x1B, 0x68, $arg0);
    return;
}

sub set_double_high {
    my ($self, ) = @_;
    $self->append_pack("CC", 0x1B, 0x0E, );
    return;
}

sub cancel_double_high {
    my ($self, ) = @_;
    $self->append_pack("CC", 0x1B, 0x14, );
    return;
}

sub set_emphasized_printing {
    my ($self, ) = @_;
    $self->append_pack("CC", 0x1B, 0x45, );
    return;
}

sub cancel_emphasized_printing {
    my ($self, ) = @_;
    $self->append_pack("CC", 0x1B, 0x46, );
    return;
}

sub select_underline_mode {
    my ($self, $arg0) = @_;
    $self->append_pack("CCC", 0x1B, 0x2D, $arg0);
    return;
}

sub select_upperline_mode {
    my ($self, $arg0) = @_;
    $self->append_pack("CCC", 0x1B, 0x5F, $arg0);
    return;
}

sub select_inverse {
    my ($self, ) = @_;
    $self->append_pack("CC", 0x1B, 0x34, );
    return;
}

sub cancel_inverse {
    my ($self, ) = @_;
    $self->append_pack("CC", 0x1B, 0x35, );
    return;
}

sub feed_n_lines {
    my ($self, $arg0) = @_;
    $self->append_pack("CCC", 0x1B, 0x61, $arg0);
    return;
}

sub set_page_length {
    my ($self, $arg0) = @_;
    $self->append_pack("CCC", 0x1B, 0x43, $arg0);
    return;
}

sub set_page_length_in_24mm_units {
    my ($self, $arg0) = @_;
    $self->append_pack("CCCC", 0x1B, 0x43, 0x00, $arg0);
    return;
}

sub set_left_margin {
    my ($self, $arg0) = @_;
    $self->append_pack("CCC", 0x1B, 0x6C, $arg0);
    return;
}

sub set_right_margin {
    my ($self, $arg0) = @_;
    $self->append_pack("CCC", 0x1B, 0x51, $arg0);
    return;
}

sub skip_lines {
    my ($self, $arg0) = @_;
    $self->append_pack("CCC", 0x1B, 0x61, $arg0);
    return;
}

sub cut {
    my ($self, ) = @_;
    $self->append_pack("CCC", 0x1B, 0x64, 0x33, );
    return;
}

sub barcode_ean13 {
    my ($self, $arg0) = @_;
    $self->append_pack("CCCCCC", 0x1B, 0x62, 0x03, 0x02, 0x01, $arg0);
    return;
}

sub barcode_code128 {
    my ($self, $arg0) = @_;
    $self->append_pack("CCCCCC", 0x1B, 0x62, 0x06, 0x02, 0x01, $arg0);
    return;
}

sub move_absolute_position {
    my ($self, $arg0, $arg1) = @_;
    $self->append_pack("CCCCC", 0x1B, 0x1D, 0x41, $arg0, $arg1);
    return;
}

sub move_relative_position {
    my ($self, $arg0, $arg1) = @_;
    $self->append_pack("CCCCC", 0x1B, 0x1D, 0x52, $arg0, $arg1);
    return;
}

sub specify_alignment {
    my ($self, $arg0) = @_;
    $self->append_pack("CCCC", 0x1B, 0x1D, 0x61, $arg0);
    return;
}

sub align_left {
    my ($self, ) = @_;
    $self->append_pack("CCCC", 0x1B, 0x1D, 0x61, 0x00, );
    return;
}

sub align_center {
    my ($self, ) = @_;
    $self->append_pack("CCCC", 0x1B, 0x1D, 0x61, 0x01, );
    return;
}

sub align_right {
    my ($self, ) = @_;
    $self->append_pack("CCCC", 0x1B, 0x1D, 0x61, 0x02, );
    return;
}

sub set_double_wide {
    my ($self, $arg0) = @_;
    $self->append_pack("CC", 0x0E, $arg0);
    return;
}

sub cancel_double_wide {
    my ($self, $arg0) = @_;
    $self->append_pack("CC", 0x14, $arg0);
    return;
}

sub select_upside_down {
    my ($self, ) = @_;
    $self->append_pack("C", 0x0F, );
    return;
}

sub cancel_upside_down {
    my ($self, ) = @_;
    $self->append_pack("C", 0x12, );
    return;
}

sub lf {
    my ($self, ) = @_;
    $self->append_pack("C", 0x0A, );
    return;
}

sub cr {
    my ($self, ) = @_;
    $self->append_pack("C", 0x13, );
    return;
}

sub crlf {
    my ($self, ) = @_;
    $self->append_pack("CC", 0x13, 0x0A, );
    return;
}

sub form_feed {
    my ($self, ) = @_;
    $self->append_pack("C", 0x0C, );
    return;
}

sub vertical_tab {
    my ($self, ) = @_;
    $self->append_pack("C", 0x0B, );
    return;
}

sub horizontal_tab {
    my ($self, ) = @_;
    $self->append_pack("C", 0x09, );
    return;
}

sub data_end {
    my ($self, ) = @_;
    $self->append_pack("C", 0x1E, );
    return;
}

sub bell {
    my ($self, ) = @_;
    $self->append_pack("C", 0x07, );
    return;
}

1;



lib/Protocol/Star/Linemode.pm
# Protocol::Star::Linemode - Generates a formatted byte string for Star POS printers
# Copyright (c) 2013 Peter Stuifzand
# Copyright (c) 2013 Other contributors as noted in the AUTHORS file
# 
# Protocol::Star::Linemode is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation; either version 3 of the License,
# or (at your option) any later version.
# 
# Protocol::Star::Linemode is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser
# General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

package Protocol::Star::Linemode;
our $VERSION = '0.1.2';
use Moo;
with qw/
Protocol::Star::Linemode::Generated
/;

has _data => (
    is      => 'rw',
    default => sub { '' },
);

sub append {
    my ($self, @data) = @_;
    $self->_data($self->_data . (join '', @data));
    return;
}

sub result {
    my $self = shift;
    return $self->_data;
}

sub append_pack {
    my ($self, $format, @args) = @_;
    $self->append(
        pack($format, @args)
    );
    return;
}

sub create_normal_code {
    my ($nargs, @code) = @_;

    my $len         = $nargs + scalar @code;
    my $pack_format = 'C' x $len;

    return sub {
        my $self = shift;
        my @args = @_;
        $self->append_pack($pack_format, @code, @args);
        return;
    }
}

sub create_escape_code {
    my ($nargs, @code) = @_;

    my $len         = 1 + $nargs + scalar @code;
    my $pack_format = 'C' x $len;

    return sub {
        my $self = shift;
        my @args = @_;
        $self->append_pack($pack_format, 0x1B, @code, @args);
        return;
    }
}

#BEGIN {
#    no strict 'refs';
#    my @escape_specs = (
#        [ 'select_font',                            1, 0x1E, 0x46 ],
#        [ 'select_code_page',                       1, 0x1D, 0x74 ],
#        [ 'set_slash_zero',                         1, 0x27 ],
#        [ 'specify_international_character_set',    1, 0x52 ],
#        [ 'specify_12_dot_pitch',                   0, 0x4D ],
#        [ 'specify_15_dot_pitch',                   0, 0x50 ],
#        [ 'specify_16_dot_pitch',                   0, 0x3A ],
#        [ 'specify_14_dot_pitch',                   0, 0x67 ],
#        [ 'set_expanded_width_height',              2, 0x69 ],
#        [ 'set_expanded_width',                     1, 0x57 ],
#        [ 'set_expanded_height',                    1, 0x68 ],
#        [ 'set_double_high',                        0, 0x0E ],
#        [ 'cancel_double_high',                     0, 0x14 ],
#        [ 'set_emphazied_printing',                 0, 0x45 ],
#        [ 'cancel_emphazied_printing',              0, 0x46 ],
#        [ 'select_underline_mode',                  1, 0x2D ],
#        [ 'select_upperline_mode',                  1, 0x5F ],
#        [ 'select_inverse',                         0, 0x34 ],
#        [ 'cancel_inverse',                         0, 0x35 ],
#        [ 'feed_n_lines',                           1, 0x61 ],
#        [ 'set_page_length',                        1, 0x43 ],
#        [ 'set_page_length_in_24mm_units',          1, 0x43, 0x00 ],
#        [ 'set_left_margin',                        1, 0x6C, ],
#        [ 'set_right_margin',                       1, 0x51, ],
#        [ 'move_absolute_position',                 2, 0x1D, 0x41 ],
#        [ 'move_relative_position',                 2, 0x1D, 0x52 ],
#        [ 'specify_alignment',                      1, 0x1D, 0x61 ],
#        [ 'align_left',                             0, 0x1D, 0x61, 0x00 ],
#        [ 'align_center',                           0, 0x1D, 0x61, 0x01 ],
#        [ 'align_right',                            0, 0x1D, 0x61, 0x02 ],
#    );
#
#    my @specs = (
#        [ 'set_double_wide',                        1, 0x0E ],
#        [ 'cancel_double_wide',                     1, 0x14 ],
#        [ 'select_upside_down',                     0, 0x0F ],
#        [ 'cancel_upside_down',                     0, 0x12 ],
#        [ 'lf',                                     0, 0x0A ],
#        [ 'cr',                                     0, 0x13 ],
#        [ 'form_feed',                              0, 0x0C ],
#        [ 'vertical_tab',                           0, 0x0B ],
#        [ 'horizontal_tab',                         0, 0x09 ],
#    );
#
#    for my $spec (@escape_specs) {
#        my $name = shift @$spec;
#        my $nargs = shift @$spec;
#        *{$name} = create_escape_code($nargs, @$spec);
#    }
#    for my $spec (@specs) {
#        my $name = shift @$spec;
#        my $nargs = shift @$spec;
#        *{$name} = create_normal_code($nargs, @$spec);
#    }
#}

sub text {
    my ($self, $text) = @_;
    $self->append_pack('A*', $text);
    return;
}

1;

=head1 NAME

Protocol::Star::Linemode - Generates a formatted byte string for Star POS printers

=head1 SYNOPSIS

  use Protocol::Star::Linemode;

  my $p = Protocol::Star::Linemode->new;
  $p->set_emphasized_printing;
  $p->text("Hello world");
  $p->cancel_emphasized_printing;

  my $formatted_output = $p->result;
  # Send $formatted_output to printer

=head1 Converting from 0.1.2 to 1.0.0

=over 4

=item * Create a L<Protocol::Star::Linemode> object instead of the L<Protocol::Star::Linemode::Generated> object

=back

=head1 AUTHOR

Peter Stuifzand E<lt>peter@stuifzand.euE<gt>

=head1 COPYRIGHT

Copyright 2013 - Peter Stuifzand

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

=cut
